### PR TITLE
refactor reader with reader registry

### PR DESCRIPTION
**Pull request by wonderflow**
## Changes
   
  - [ ] refactor reader with reader registry
  - [ ] fix isFromWeb to errDIrectReturn  
  - [ ] add csv not match case
  - [ ] support mutiline json
 
## Reviewers

  - [ ] @redHJ  please review

## Wiki Changes
 
  - options1...
  - options2...
    
## Checklist
   
   - [ ] Rebased/mergable
   - [ ] Tests pass
   - [ ] Wiki updated

### CODE REVIEW

File: mgr/dataflow_test.go [hunk 1]
```diff
@@ -245,7 +245,8 @@ func Test_SendData(t *testing.T) {
 	meta, err := reader.NewMetaWithConf(readConf)
 	assert.NoError(t, err)
 	defer os.RemoveAll("./meta")
-	httpReader, err := reader.NewHttpReader(meta, c)
+	hhttpReader, err := reader.NewHttpReader(meta, c)
+	httpReader :=hhttpReader.(*reader.HttpReader)
 	assert.NoError(t, err)
```
Comment (redHJ) [inline mgr/dataflow_test.go]: assert.NoError(t, err)应该放在httpReader :=hhttpReader.(*reader.HttpReader)之前，如果hhttpReader是nil会出错？

File: mgr/mgr.go [hunk 2]
```diff
@@ -22,6 +22,7 @@ import (
 
 	"github.com/howeyc/fsnotify"
 	"github.com/json-iterator/go"
+	"github.com/qiniu/logkit/reader"
```
Comment (redHJ) [inline mgr/mgr.go]: 顺序

File: mgr/mgr.go [hunk 3]
```diff
@@ -98,6 +101,7 @@ func NewCustomManager(conf ManagerConfig, pr *parser.ParserRegistry, sr *sender.
 		watchers:      make(map[string]*fsnotify.Watcher),
 		pregistry:     pr,
 		sregistry:     sr,
+		rregistry:rr,
```
Comment (redHJ) [inline mgr/mgr.go]: format

File: mgr/dataflow_test.go (hunk 1 above)
Comment (wonderflow) [reply mgr/dataflow_test.go]: 都是测试，问题不大

File: mgr/rest_test.go [hunk 4]
```diff
@@ -22,6 +22,8 @@ import (
 	"github.com/qiniu/logkit/reader"
 	. "github.com/qiniu/logkit/utils/models"
 
+	"sync"
```
Comment (redHJ) [inline mgr/rest_test.go]: format

File: mgr/dataflow_test.go [hunk 5]
```diff
@@ -245,7 +245,8 @@ func Test_SendData(t *testing.T) {
 	meta, err := reader.NewMetaWithConf(readConf)
 	assert.NoError(t, err)
 	defer os.RemoveAll("./meta")
-	httpReader, err := reader.NewHttpReader(meta, c)
+	hhttpReader, err := reader.NewHttpReader(meta, c)
+	httpReader := hhttpReader.(*reader.HttpReader)
 	assert.NoError(t, err)
```
Comment (redHJ) [inline mgr/dataflow_test.go]: err放前面

File: generators/grok_pattern_generater.go [hunk 6]
```diff
@@ -28,7 +28,7 @@ func main() {
 			data, err := rd.ReadBytes('\n')
 			if err != nil {
 				if err != io.EOF {
-					fmt.Println("ERROR:", err)
+					log.Error("ERROR:", err)
```
Comment (redHJ) [inline generators/grok_pattern_generater.go]: log打详细一点？

File: reader/bufreader.go [hunk 7]
```diff
@@ -18,6 +18,8 @@ import (
 	"sync/atomic"
 	"unsafe"
 
+	"time"
+
```
Comment (redHJ) [inline reader/bufreader.go]: format

File: reader/http_reader_test.go [hunk 8]
```diff
@@ -28,7 +28,8 @@ func TestNewHttpReader(t *testing.T) {
 	meta, err := NewMetaWithConf(readConf)
 	assert.NoError(t, err)
 	defer os.RemoveAll("./meta")
-	httpReader, err := NewHttpReader(meta, c)
+	hhttpReader, err := NewHttpReader(meta, c)
+	httpReader := hhttpReader.(*HttpReader)
 	assert.NoError(t, err)
```
Comment (redHJ) [inline reader/http_reader_test.go]: httpReader := hhttpReader.(*HttpReader)
assert.NoError(t, err)
换个顺序

**Pull request merged**